### PR TITLE
Quartz beta dev screenshot

### PR DESCRIPTION
**Pull request by manishk3189**
The sample demonstrates how to export the map to an image [#983](https://devtopia.esri.com/runtime/java-android-sdk/issues/983) 

@doneill Can you please review the PR and let me know of changes if needed.

### CODE REVIEW

File: take-screenshot/src/main/java/com/esri/arcgisruntime/sample/takescreenshot/MainActivity.java [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+    /**
+     * save the bitmap image to file and open it
+     *
+     * @param bitmap
+     * @throws IOException
+     */
+    private File saveToFile(Bitmap bitmap) throws IOException {
+
+        // create a directory ArcGIS to save the file
+        File root = null;
```
Comment (doneill) [inline take-screenshot/src/main/java/com/esri/arcgisruntime/sample/takescreenshot/MainActivity.java]: Initializing as null is redundant, remove redundancy.

File: take-screenshot/src/main/java/com/esri/arcgisruntime/sample/takescreenshot/MainActivity.java [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+        String fileName = "map-export-image" + System.currentTimeMillis() + ".png";
+        root = Environment.getExternalStorageDirectory();
+        File fileDir = new File(root.getAbsolutePath() + "/ArcGIS/");
+        fileDir.mkdirs();
```
Comment (doneill) [inline take-screenshot/src/main/java/com/esri/arcgisruntime/sample/takescreenshot/MainActivity.java]: You are ignoring the result of `mkdirs()`.  Please test the result and act accordingly.  Something like below:

```java
 boolean isDirectoryCreated=fileDir.exists();
 if (!isDirectoryCreated) {
     isDirectoryCreated= fileDir.mkdirs();
 }
 if(isDirectoryCreated) {
    // do something
 }
```

File: take-screenshot/src/main/java/com/esri/arcgisruntime/sample/takescreenshot/MainActivity.java [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+        fos.close();
+
+        return file;
+
+    }
+
+    /**
+     * AsyncTask class to save the bitmap as an image
+     */
+    public class SaveImageTask extends AsyncTask<Bitmap, Void, File> {
```
Comment (doneill) [inline take-screenshot/src/main/java/com/esri/arcgisruntime/sample/takescreenshot/MainActivity.java]: Declaration should be private.

File: take-screenshot/src/main/java/com/esri/arcgisruntime/sample/takescreenshot/MainActivity.java [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+import com.esri.arcgisruntime.mapping.view.MapView;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+
+public class MainActivity extends AppCompatActivity {
+
+    private MapView mMapView;
+    private String TAG = "TakeScreenshot";
```
Comment (doneill) [inline take-screenshot/src/main/java/com/esri/arcgisruntime/sample/takescreenshot/MainActivity.java]: Declaration should be final.

File: take-screenshot/src/main/res/values/dimens.xml [hunk 5]
```diff
@@ -0,0 +1,6 @@
+<resources>
+    <!-- Default screen margins, per the Android Design guidelines. -->
+    <dimen name="activity_horizontal_margin">16dp</dimen>
+    <dimen name="activity_vertical_margin">16dp</dimen>
+    <dimen name="fab_margin">16dp</dimen>
```
Comment (doneill) [inline take-screenshot/src/main/res/values/dimens.xml]: This resource is unused, ok to remove?

File: take-screenshot/build.gradle [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+    buildTypes {
+        release {
+            minifyEnabled false
+            proguardFiles getDefaultProguardFile('proguard-android.txt'), 'proguard-rules.pro'
+        }
+    }
+}
+
+dependencies {
+    compile 'com.android.support:appcompat-v7:23.2.1'
```
Comment (doneill) [inline take-screenshot/build.gradle]: Update and test against v24.0.0

File: take-screenshot/build.gradle [hunk 7]
```diff
@@ -0,0 +1,25 @@
+apply plugin: 'com.android.application'
+
+android {
+    compileSdkVersion rootProject.ext.compileSdkVersion
+    buildToolsVersion rootProject.ext.buildToolsVersion
+
+    defaultConfig {
+        applicationId "com.esri.arcgisruntime.sample.takescreenshot"
+        minSdkVersion 16
+        targetSdkVersion 23
```
Comment (doneill) [inline take-screenshot/build.gradle]: Update and test against targetSdkVersion 24

File: take-screenshot/src/main/AndroidManifest.xml [hunk 8]
```diff
@@ -0,0 +1,25 @@
+<?xml version="1.0" encoding="utf-8"?>
+<manifest xmlns:android="http://schemas.android.com/apk/res/android"
+    package="com.esri.arcgisruntime.sample.takescreenshot">
+
+    <uses-permission android:name="android.permission.INTERNET" />
+    <uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE" />
+    <uses-permission android:name="android.permission.READ_EXTERNAL_STORAGE" />
```
Comment (doneill) [inline take-screenshot/src/main/AndroidManifest.xml]: `READ_EXTERNAL_STORAGE` is redundant as `WRITE_EXTERNAL_STORAGE` includes read/write permission.

File: take-screenshot/src/main/java/com/esri/arcgisruntime/sample/takescreenshot/MainActivity.java [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+        String fileName = "map-export-image" + System.currentTimeMillis() + ".png";
+        root = Environment.getExternalStorageDirectory();
+        File fileDir = new File(root.getAbsolutePath() + "/ArcGIS/");
+        fileDir.mkdirs();
+
+        // create the file inside the directory
+        File file = new File(fileDir, fileName);
+
+        // write the bitmap to PNG file
+        FileOutputStream fos = new FileOutputStream(file);
```
Comment (doneill) [inline take-screenshot/src/main/java/com/esri/arcgisruntime/sample/takescreenshot/MainActivity.java]: This is causing `File` to be null and crashing the app with a `NullPointerException` on line 182. Commenting the `FileOutputStream` avoids the crash.  Please investigate the cause and determine if this is a required fix or whether we can return the `File` as is.

**Comment by doneill** [comment]
@manishk3189 I gave a first pass at reviewing your code, please address comments with special attention to the app crashes when attempting to take a screenshot with a `NullPointerException` as `file=null` on line 183.

**Comment by manishk3189** [comment]
@doneill The development of the sample was done with targetSdk 22 and was working fine. During development, I upgraded the target version to 23 but it didn't throw runtime exception as app was upgraded on same device. The app was crashing because the sample didn't had required permissions at runtime and hence was unable to make directory or save image as a file. 

I have refactored the sample to include runtime permissions and directory checks. Can you please review the same and let me know of any changes if required?

**Comment by doneill** [comment]
@mani8177 Updates look good, will merge.

**Pull request merged**